neonUtilities/R/zipsByProduct.R

```diff
@@ -338,27 +196,52 @@ zipsByProduct <- function(dpID, site="all", startdate=NA, enddate=NA, package="b
   }
   dir.create(filepath)
 
-  writeLines(paste("Downloading ", nrow(zip.urls)-1, " files", sep=""))
+  writeLines(paste("Downloading ", nrow(zip.urls), " files", sep=""))
   pb <- utils::txtProgressBar(style=3)
   utils::setTxtProgressBar(pb, 1/(nrow(zip.urls)-1))
-  # copy zip files into folder
-  for(i in 2:nrow(zip.urls)) {
-    zip_out <- paste(filepath, zip.urls$name[i], sep="/")
-    if(!file.exists(substr(zip_out, 1, nchar(zip_out)-4)) || !file.exists(zip_out)) {
-      downloader::download(zip.urls$URL[i], zip_out, 
-                           mode="wb", quiet=T)
-      utils::setTxtProgressBar(pb, i/(nrow(zip.urls)-1))
+
+  j <- 1
+  counter<- 1
+
+  while(j <= nrow(zip.urls)) {
+    counter <- counter + 1
+
+    if (counter > 3) {
+      cat(paste0("\nURL query ", zip.urls$name[j],
+                  " failed. The API or data product requested may be unavailable at this time; check data portal (data.neonscience.org/news) for possible outage alert."))
+      j <- j + 1
+    } else {
+      zip_out <- paste(filepath, zip.urls$name[j], sep="/")
+      if(!file.exists(substr(zip_out, 1, nchar(zip_out)-4)) || !file.exists(zip_out)) {
+        t <- tryCatch(
+          {
+            downloader::download(zip.urls$URL[j], zip_out,
+                                 mode="wb", quiet=T)
+          }, error = function(e) { e } )
+
+        if(inherits(t, "error")) {
+          writeLines("File could not be downloaded. URLs may have expired. Trying new URLs.")
+          zip.urls <- getZipUrls(tmp.files, month.urls=month.urls, avg=avg, package=package,
+                                 max.pub=max.pub, max.pub.site=max.pub.site, dpID=dpID, messages=messages) %>%
+            tidyr::drop_na()
```

Comment:
Something isn't working quite right here. When I test this, it returns the 'trying new URLs' message, but then fails for all remaining files. It also prints the warnings from the initial attempt. In the AOP download functions, the download() call is wrapped in suppressWarnings(), so that should be added here.

I think the failures for subsequent URLs are happening because in the error case, the counter never gets set back to 1. I think you need to add `counter <- 1 ` at the end of the `if(counter>3)` conditional.

Given that, I'm not sure why this worked correctly for AOP.  It seems like the same problem should have happened there. So it's possible I've misdiagnosed, take a look and see if my proposed solution works.

The test cases I've been using are 

```
zipsByProduct(dpID='DP4.00200.001', site='all', check.size=F, 
          savepath='/Users/clunch/Desktop')
```

and 

```
byFileAOP(dpID='DP1.30006.001', site='HARV', year=2017, check.size=F, 
          savepath='/Users/clunch/Desktop')
```

Each of these takes a little more than an hour on my machine, on the internal network.

Reply:
@cklunch I think you nailed the diagnosis. Made those 2 changes and added the counter <-1 in the 2 AOP functions